Packages/vcs/Lib/colorpicker.py

```diff
@@ -0,0 +1,212 @@
+        self.selectedActor.GetProperty().SetLineWidth(3);
+        self.color = color
+        # Make sure the current color is selected
+        self.selectCell(color)
+        self.color_renderer.AddActor(self.selectedActor)
+        self.click_handler = inter.AddObserver("LeftButtonPressEvent", self.leftButtonPressEvent)
+
+        inter.Start()
+
+    def make_current(self):
+        self.render_window.MakeCurrent()
+
+    def topRendererAtPoint(self, x, y):
+        inter = self.render_window.GetInteractor()
+
+        top = None
+
+        for renderer in collection(inter.GetRenderWindow().GetRenderers()):
+            if renderer.IsInViewport(x, y):
+                if top is None or top.GetLayer() <= renderer.GetLayer():
+                    top = renderer
+        return top
+
+    def change_map(self, state):
+        self.colormap = self.colormaps[state]
+        self.colors =  [[int(c / 100.0 * 255.0) for c in self.colormap.index[i]] for i in range(len(self.colormap.index))]
+        colorData = colors_to_scalars(self.colors)
+        self.actor.GetMapper().GetInput().GetCellData().SetScalars(colorData)
+        self.actor.GetMapper().Update()
+        self.render_window.Render()
+
+    def save(self, state):
+        if self.on_save is not None:
+            self.on_save(self.colormap, self.color)
+
+        self.close()
+
+    def cancel(self, state):
+        if self.on_cancel is not None:
+            self.on_cancel()
+
+        self.close()
+
+    def selectCell(self, cellId):
+        if cellId == -1:
+            return
+        ids = vtk.vtkIdTypeArray();
+        ids.SetNumberOfComponents(1);
+        ids.InsertNextValue(cellId);
+
+        selectionNode = vtk.vtkSelectionNode();
+        selectionNode.SetFieldType(vtk.vtkSelectionNode.CELL);
+        selectionNode.SetContentType(vtk.vtkSelectionNode.INDICES);
+        selectionNode.SetSelectionList(ids);
+
+        selection = vtk.vtkSelection();
+        selection.AddNode(selectionNode);
+
+        extractSelection = vtk.vtkExtractSelection();
+
+        extractSelection.SetInputData(0, self.actor.GetMapper().GetInput());
+        extractSelection.SetInputData(1, selection);
+
+        extractSelection.Update();
+
+        selected = vtk.vtkUnstructuredGrid();
+        selected.ShallowCopy(extractSelection.GetOutput());
+
+        self.selectedMapper.SetInputData(selected);
+        self.selectedMapper.Update()
+
+    def leftButtonPressEvent(self, obj, event):
+        inter = self.render_window.GetInteractor()
+
+        x, y = inter.GetEventPosition()
+
+        renderer = self.topRendererAtPoint(x, y)
+
+        if renderer:
+            picker = vtk.vtkCellPicker()
+            picker.SetTolerance(.0005)
+            picker.Pick(x, y, 0, renderer)
+
+            cell = picker.GetCellId()
+            self.selectCell(cell)
+            if cell >= 0:
+                self.color = int(cell)
+
+            self.render_window.Render()
+
+
+    def close(self):
```

Comment:
I dug around a little bit, and found out the following:

- `TerminateApp` appears to do what it says, and will terminate the application (at least in the Cocoa implementation)
- `vtkRenderWindowInteractor` will call `TerminateApp` if there is not an `ExitEvent` observer set when the `ExitCallback` is called.
- `vtkCocoaRenderWindowInteractor` overrides this behavior to call `TerminateApp` no matter what when `ExitCallback` is triggered.
- `vtkInteractorStyle`, by default, will invoke `ExitCallback` on the interactor whenever `[Q|q|E|e]` are pressed. This is in the `OnChar` handler.

So with the default interactor style, `TerminateApp` will always be called when 'q' or 'e' are pressed.

This sounds consistent with what you're seeing -- pressing q/e in the colorpicker is triggering the app to terminate unexpectly, correct?

I'd suggest trying a `vtkInteractorStyleUser` style, which allows you to override the default event handlers. It will still call the default `OnChar` method (which in turn will call `ExitCallback` and `TerminateApp`) if there is not a `CharEvent` observer set on it. So even if you aren't handling keystrokes, you'll want to make sure a no-op observer is set on the interactor style object's `CharEvent'.

Reading this back, it's pretty convoluted. Let me know if I need to clear anything up ;)